File: LeetCode-c++/46_permute.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution
{
    public:
    vector<vector<int>> result;
    void backtrack(vector<bool>& used, vector<int>& nums, vector<int>& path)
    {
        if(path.size() == nums.size())
        {
            result.push_back(path);
            return;
        }
        for(int i = 0; i<nums.size(); ++i)
        {
            if(!used[i])
            {    
                path.push_back(nums[i]);
                used[i] = true;
                backtrack(used, nums, path);
                path.pop_back();
                used[i] =false;
            }
        }

    }
    vector<vector<int>> permute(vector<int>& nums)
    {
        vector<bool> used = vector<bool>(nums.size(), false);
        vector<int> path;
        backtrack(used,  nums, path);
        return result;
    }
};

int main()
{
    vector<int> nums = {1, 2, 3};
    Solution solution = Solution();
    vector<vector<int>> res = solution.permute(nums);
    for(vector<int> x: res)
    {
        cout << '[';
        for(int i: x)
        {
            cout << i << ' ';
        }
        cout << ']' << endl;
    }
    system("pause");
    return 0;
}

File: LeetCode-py/46_permute.py
class Solution:
    @classmethod
    def permute(self, nums):
        result = []
        used = [False for _ in range(len(nums))]
        def backtrack(path):
            if(len(path) == len(nums)):
                result.append(path)
                return
            for i in range(len(nums)):
                if not used[i]:
                    used[i] = True
                    backtrack(path + [nums[i]])
                    used[i] = False
        backtrack([])
        return result

if __name__ == '__main__':
    nums = [1, 2, 3]
    print(Solution.permute(nums))
